pkg/jobWorker/job.go

```diff
@@ -0,0 +1,374 @@
+// The user running Start() should be the root user or have the necessary permissions to create namespaces and control groups
+//
+// ErrJobAlreadyStarted is returned, if the Job has already been started.
+// ErrInvalidCommand, ErrInvalidCPU, ErrInvalidIOBytesPerSecond, ErrInvalidMemBytes is returned, if provided configuration is invalid
+func (job *Job) Start() error {
+	job.mutex.Lock()
+	defer job.mutex.Unlock()
+
+	// validate configuration
+	log.Printf("validate job:%s", job)
+	if err := job.config.isValid(); err != nil {
+		log.Printf("validate job error:%v", err)
+		return err
+	}
+
+	// validate job hasn't been run
+	if job.isStarted {
+		return ErrJobAlreadyStarted
+	}
+
+	cmd := exec.Command(job.config.Command, job.config.Arguments...)
+	// combine the stdout and stderr so that the stdout and stderr are combined in the order they are written
+	cmd.Stderr = job.output
+	cmd.Stdout = job.output
+
+	cmd.SysProcAttr = &syscall.SysProcAttr{
+		// CLONE_NEWPID:  creates a new PID namespace preventing the process from seeing/killing host processes
+		// CLONE_NEWNET:  creates a new network namespace preventing the process from accessing the internet or local network
+		// CLONE_NEWNS:   creates a new mount namespace preventing the process from impacting host mounts
+		// CLONE_NEWUTS:  creates a new UTS namespaces provide isolation between two system identifiers: the hostname and the NIS domain name
+		// CLONE_NEWPID:  crates new PID namespaces isolate the process ID number space, meaning that processes in different PID namespaces can have the same PID
+		// CLONE_NEWUSER: creates new namespaces to isolate security-related identifiers and attributes, in particular, user IDs and group IDs
+		Cloneflags: syscall.CLONE_NEWNS |
+			syscall.CLONE_NEWIPC |
+			syscall.CLONE_NEWNET |
+			syscall.CLONE_NEWUTS |
+			syscall.CLONE_NEWPID |
+			syscall.CLONE_NEWUSER,
+		UidMappings: []syscall.SysProcIDMap{
+			{
+				ContainerID: 0,
+				HostID:      os.Getuid(),
+				Size:        1,
+			},
+		},
+		GidMappings: []syscall.SysProcIDMap{
+			{
+				ContainerID: 0,
+				HostID:      os.Getgid(),
+				Size:        1,
+			},
+		},
+		// Also, enables mounting a new proc filesystem so that command such as `ps -ef` only see the processes in the PID namespace
+		Unshareflags: syscall.CLONE_NEWNS,
+		// instruct cmd.Run to use the control group file descriptor, so that Job Command does not
+		// have to manually add the new PID to the control group
+		UseCgroupFD: true,
+	}
+
+	cleanCGroup := make(chan bool, 1)
+	go func() {
+		select {
+		case <-cleanCGroup:
+			{
+				// do not close the cgroup.procs file until after the process has exited
+				if err := ns.DeleteCGroup(job.getCGroupName()); err != nil {
+					log.Printf("error closing cgroup: %s\n", err)
+					job.exitReason = errors.Join(job.exitReason, fmt.Errorf("error closing cgroup: %w\n", err))
```

Comment:
@smallinsky sorry, trying to understand your point about `can be scheduled after the function exit` 

I think I can create some implementation where waiting `go func()` starting after `Start' function, but in any case `go func()` waiting when 'Start' release 'Lock' to continue. See on the picture below.
<img width="897" alt="Screenshot 2024-09-05 at 1 17 58 PM" src="https://github.com/user-attachments/assets/9ca7cc9c-880c-4008-897c-a9a43a513218">

And I run intentionally `processState, err := job.cmd.Process.Wait()` before 'job.mutex.Lock()' to prevent lock on other functions and left them execute freely  
 
```go
		processState, err := job.cmd.Process.Wait()
		job.processState = processState

		job.mutex.Lock()
		defer job.mutex.Unlock()
```

Is that ok ?

Reply:
>sorry, trying to understand your point about can be scheduled after the function exit

Sorry for confusion.

The question is: what lock protects the following job.exitReason write [operation](https://github.com/P-A-R-U-S/Go-Job-Worker-Service/pull/2/files/51b5bbd76770e793822084936875536f669a2fb3#diff-35dec03abce473a482ac214d8104767d523b60d5cdf7d1248f7e02e24a8ccd6aR194)?

```go
if err := ns.DeleteCGroup(job.getCGroupName()); err != nil {
			log.Printf("error closing cgroup: %s\n", err)
	      job.exitReason = errors.Join(job.exitReason, fmt.Errorf("error closing cgroup: %w\n", err))
}
```

The current implementation assumes that the lock acquired by:
```go
  processState, err := job.cmd.Process.Wait()
  job.mutex.Lock()
  defer job.mutex.Unlock()
  ```
protects the job.exitReason write, but the lock can be released earlier, and the following:

```
job.exitReason = errors.Join(job.exitReason, fmt.Errorf("error closing cgroup: %w\n", err))
```
could be written after the lock is released.

Imagine following scenarion where the gorutine
```
  defer func() {
    cleanCGroup <- true
    unmount <- true
  }()
  ```

is delayed by GO  scheduler and in meantime a the job.Status() is called where exitReason is read under read lock `job.getExitReason(),`  

```go
func Start() {
    go func() {
        select {
        case <-unmount:
            if err := ns.UnmountProc(); err != nil {
                log.Printf("error unmounting /proc - %s\n", err)
                job.exitReason = errors.Join(job.exitReason, fmt.Errorf("error unmounting /proc - %w\n", err))
            }
        }
    }()

    go func() {
        processState, err := job.cmd.Process.Wait()
        job.mutex.Lock()
        defer job.mutex.Unlock()
        job.isCompleted = true

        defer func() {
            sleep(time.Second * 10)
            cleanCGroup <- true
            unmount <- true
        }()
        
        if err = job.output.Close(); err != nil {
            job.exitReason = errors.Join(job.exitReason, fmt.Errorf("error closing output: %w\n", err))
        }
        return
    }()
}

```

The current flow doesn't quarantine the that `job.exitReason = errors.Join(job.exitReason, fmt.Errorf("error closing output: %w\n", err))` is called under the lock

Reply:
@smallinsky Thank you Marek for details - now I got your point. As I see there are two possible options to resolved it.

Using `biderectional channel` and add defer functions to wait cleanup status. Because defer execute in LIFO order - all of them would be executed before we actually release Lock.

![Screenshot 2024-09-06 at 1 15 34 PM](https://github.com/user-attachments/assets/9ec34799-3bf9-4846-9392-a18a6e2a8cfc)

And it is quite tricky to design unit test to check this, but I've `artificially` slow down go-routines which `deleting cgroups/unmount`, added log and here is executions process as a proof. 

![Screenshot 2024-09-06 at 1 33 31 PM](https://github.com/user-attachments/assets/5d89d145-4307-4d0e-a2a1-a1667b13c8e4)


2: Other possible options using watGroup, but looks like we end-up with case when `Start` functions blocked and waiting until `cgroups/mount` clean-up ether from`Start` function or from `go func()`. And some other refactoring probably need. As for me it is looks complicated and potentially introduce extra effort to support it in future.  

Let me know what do you think ?

Reply:
> And it is quite tricky to design unit test to check this, but I've artificially slow down go-routines which

https://github.com/P-A-R-U-S/Go-Job-Worker-Service/pull/2/files#diff-35dec03abce473a482ac214d8104767d523b60d5cdf7d1248f7e02e24a8ccd6aR203

This seem to be very hack way to test things caused by overcomplicated code. 

If you need to call 
```go
 if err {
    cleanCGroup <- true
     waitingCleanCGroupToCompleted 
	  defer waitingCleanCGroupToCompleted()
		return fmt.Errorf("error creating cgroup: %w", err)
}
```
why not just  move cgrup into function and do the cleanup synchronous  ?

Reply:
Done. Agree code looks much easy.